Source/WebKit/NetworkProcess/storage/BackgroundFetchStoreManager.cpp

```diff
@@ -0,0 +1,217 @@
+            auto fileSize = FileSystem::fileSize(fetchPath);
+            if (!fileSize)
+                continue;
+
+            auto fetchData = SharedBuffer::createWithContentsOfFile(fetchPath, FileSystem::MappedFileMode::Private, shouldUseFileMapping(*fileSize) ? SharedBuffer::MayUseFileMapping::Yes : SharedBuffer::MayUseFileMapping::No);
+            fetches.append(std::pair(WTFMove(fetchData), WTFMove(fileName)));
+        }
+
+        queue->dispatch([callback = WTFMove(callback), fetches = WTFMove(fetches)]() mutable {
+            callback(WTFMove(fetches));
+        });
+    });
+}
+
+void BackgroundFetchStoreManager::clearFetch(const String& identifier, CompletionHandler<void()>&& callback)
+{
+    assertIsCurrent(m_taskQueue.get());
+
+    if (m_path.isEmpty()) {
+        callback();
+        return;
+    }
+
+    auto filePath = FileSystem::pathByAppendingComponents(m_path, { identifier });
+    m_ioQueue->dispatch([queue = Ref { m_taskQueue }, filePath = WTFMove(filePath).isolatedCopy(), callback = WTFMove(callback)]() mutable {
+        FileSystem::deleteFile(filePath);
+        size_t index = 0;
+        String bodyPath;
+        do {
+            bodyPath = makeString(filePath, "-", index++);
+        } while (FileSystem::deleteFile(bodyPath));
+        queue->dispatch(WTFMove(callback));
+    });
+}
+
+void BackgroundFetchStoreManager::clearAllFetches(const Vector<String>& identifiers, CompletionHandler<void()>&& callback)
+{
+    assertIsCurrent(m_taskQueue.get());
+
+    if (m_path.isEmpty()) {
+        callback();
+        return;
+    }
+
+    auto filePaths = map(identifiers, [this](auto& identifier) -> String {
+        return FileSystem::pathByAppendingComponents(m_path, { identifier });
+    });
+    m_ioQueue->dispatch([queue = Ref { m_taskQueue }, filePaths = crossThreadCopy(WTFMove(filePaths)), callback = WTFMove(callback)]() mutable {
+        for (auto& filePath : filePaths) {
+            FileSystem::deleteFile(filePath);
+
+            size_t index = 0;
+            String bodyPath;
+            do {
+                bodyPath = makeString(filePath, "-", index++);
+            } while (FileSystem::deleteFile(bodyPath));
+        }
+        queue->dispatch(WTFMove(callback));
+    });
+}
+
+void BackgroundFetchStoreManager::storeFetch(const String& identifier, Vector<uint8_t>&& data, CompletionHandler<void(StoreResult)>&& callback)
+{
+    assertIsCurrent(m_taskQueue.get());
+
+    if (m_path.isEmpty()) {
+        callback(StoreResult::OK);
+        return;
+    }
+
+    auto filePath = FileSystem::pathByAppendingComponents(m_path, { identifier });
+    m_ioQueue->dispatch([queue = Ref { m_taskQueue }, filePath = WTFMove(filePath).isolatedCopy(), data = WTFMove(data), callback = WTFMove(callback)]() mutable {
+        auto writeResult = FileSystem::overwriteEntireFile(filePath, { data.data(), data.size() });
+        auto result = writeResult != -1 ? StoreResult::OK : StoreResult::InternalError;
```

Comment:
Should write size < buffer size be counted as failure?

Reply:
Probably, I'll add a FIXME like done below for body chunks